File: src/main/R/bikeshare.R
library(caret)
library(lubridate)
## set seed
set.seed(32343)
data <- read.csv("~/Datasets/kaggle/bike-sharing-demand/train.csv", stringsAsFactors = FALSE)
data$wday <- wday(data$datetime)
data$hour <- hour(data$datetime)

predictors <- colnames(data)
# predictors <- predictors[colSums(is.na(data)) == 0]
predictors <- predictors[c(-1, -10, -11)]
nsv <- nearZeroVar(data[, predictors])
predictors <- predictors[-nsv]


inTrain <- createDataPartition(y=data$count,
                               p=0.7, list=FALSE)
training <- data[inTrain, predictors]
testing <- data[-inTrain, predictors]

modellm <- train(count ~ ., data=training, method="lm")
predictlm <- predict(modellm, newdata=testing)
rmselm <- sqrt(mean((predictlm - testing$count) ^ 2, na.rm = TRUE))

modelrf <- train(count ~ ., data=training, method="rf", ntree = 100)
predictrf <- predict(modelrf, newdata=testing)
rmserf <- sqrt(mean((predictrf - testing$count) ^ 2, na.rm = TRUE))

# submission no. 1.
training <- data[, predictors]
modellm <- train(count ~ ., data=training, method="lm")

test <- read.csv("~/Datasets/kaggle/bike-sharing-demand/test.csv", stringsAsFactors = FALSE)
test$wday <- wday(test$datetime)
test$hour <- hour(test$datetime)
testing <- test[, colnames(test) %in% predictors]

predictlm <- predict(modellm, newdata=testing)

# remove negative
predictlm <- pmax(predictlm, 0)

write.csv(x = data.frame(datetime = test$datetime, count = as.integer(round(predictlm))), file = "bikeshare_submission_01.csv", row.names = FALSE)


# submission no. 2.
training <- data[, predictors]
modelrf <- train(count ~ ., data=training, method="rf", ntree = 100)

test <- read.csv("~/Datasets/kaggle/bike-sharing-demand/test.csv", stringsAsFactors = FALSE)
test$wday <- wday(test$datetime)
test$hour <- hour(test$datetime)
testing <- test[, colnames(test) %in% predictors]

predictrf <- predict(modelrf, newdata=testing)

# remove negative
predictrf <- pmax(predictrf, 0)

write.csv(x = data.frame(datetime = test$datetime, count = as.integer(round(predictrf))), file = "bikeshare_submission_02.csv", row.names = FALSE)

File: src/main/python/bikeshare.py
import pandas as pd
import numpy as np
import time

def pre_process(df, train = True):
    if (train):
        df = df.drop(['casual', 'registered', 'holiday'], axis=1)
    else:
         df = df.drop(['holiday'], axis=1)
    df['datetime'] = pd.to_datetime(df['datetime'])
    df.set_index('datetime', inplace=True)
    df['wday'] = df.index.weekday
    df['hour'] = df.index.hour
    
    if (train):
        return df.drop(['count'], axis=1).values, df['count'].values
    else:
        return df.values

# For .read_csv, always use header=0 when you know row 0 is the header row
train_df = pd.read_csv('/home/namukhtar/Datasets/kaggle/bike-sharing-demand/train.csv', header=0)
# print train_df.head(10)
train_data, label = pre_process(train_df)

test_df = pd.read_csv('/home/namukhtar/Datasets/kaggle/bike-sharing-demand/test.csv', header=0)
# print test_df.head(10)
test_data = pre_process(test_df, train = False)

# Import the random forest package
from sklearn import linear_model

regr = linear_model.LinearRegression()

regr = regr.fit(train_data, label)

# Take the same decision trees and run it on the test data
output = regr.predict(test_data)

out_df = pd.DataFrame({'datetime' : test_df['datetime']})
out_df["count"] = np.round(np.maximum(output,0), decimals=0).astype('int')

out_df.to_csv('bikeshare-linear-regression.csv', index=False)
